trie-db/src/nibble/mod.rs

```
@@ -0,0 +1,192 @@
+	pub const NIBBLE_PER_BYTE : usize = 2;
+	/// Number of child for a branch (trie radix).
+	pub const NIBBLE_LENGTH : usize = 16;
+	/// Nibble (half a byte).
+	pub const PADDING_BITMASK: u8 = 0x0F;
+	/// Size of header.
+	pub const CONTENT_HEADER_SIZE: u8 = 1;
+
+	/// Mask a byte, keeping left nibble.
+	#[inline(always)]
+	pub fn pad_left(b: u8) -> u8 {
+		b & !PADDING_BITMASK
+	}
+
+	/// Mask a byte, keeping right byte.
+	#[inline(always)]
+	pub fn pad_right(b: u8) -> u8 {
+		b & PADDING_BITMASK
+	}
+
+	/// Get u8 nibble value at a given index of a byte.
+	#[inline(always)]
+	pub fn at_left(ix: u8, b: u8) -> u8 {
+		if ix == 1 {
+			b & PADDING_BITMASK
+		} else {
+			b >> BIT_PER_NIBBLE
+		}
+	}
+
+	/// Get u8 nibble value at a given index in a left aligned array.
+	#[inline(always)]
+	pub fn left_nibble_at(v1: &[u8], ix: usize) -> u8 {
+		at_left(
+			(ix % NIBBLE_PER_BYTE) as u8,
+			v1[ix / NIBBLE_PER_BYTE]
+		)
+	}
+
+	/// Get u8 nibble value at a given index in a `NibbleSlice`.
+	#[inline(always)]
+	pub fn at(s: &NibbleSlice, i: usize) -> u8 {
+		let ix = (s.offset + i) / NIBBLE_PER_BYTE;
+		let pad = (s.offset + i) % NIBBLE_PER_BYTE;
+		at_left(pad as u8, s.data[ix])
+	}
+
+	/// Push u8 nibble value at a given index into an existing byte.
+	#[inline(always)]
+	pub fn push_at_left(ix: u8, v: u8, into: u8) -> u8 {
+		into | if ix == 1 {
+			v
+		} else {
+			v << BIT_PER_NIBBLE
+		}
+	}
+
+	#[inline]
+	/// Calculate the number of needed padding a array of nibble length `i`.
+	pub fn number_padding(i: usize) -> usize {
+		i % NIBBLE_PER_BYTE
+	}
+
+	/// The nibble shifts needed to align.
+	/// We use two value, one is a left shift and
+	/// the other is a right shift.
+	pub const SPLIT_SHIFTS: (usize, usize) = (4, 4);
+
+	/// Count the biggest common depth between two left aligned packed nibble slice.
+	pub fn biggest_depth(v1: &[u8], v2: &[u8]) -> usize {
+		// sorted assertion preventing out of bound
+		for a in 0..v1.len() {
+			if v1[a] == v2[a] {
```

Is that always correct that `v1.len() <= v2.len()` (otherwise it'll panic)? I have found the only usage of the `biggest_depth()` from the `trie_visit()`. I assume that `input` there is ordered in some manner? If it sorted using `BTreeMap`, then given that e.g. `vec![5, ,5] < vec![20]`, this could panic, no?

---

In fact there is no verification of the ordering, iter_build code does not support unordered input.
I can prevent this failing, but I will need to propagate the error, that seems pretty quick to do.

---

Sorry - I mean that even if input is ordered (lexicographically, I suppose), then this could panic, because lexicographically vec of larger len could came before vec of smaller len. No need to propagate error imo.

---

yes (I just realized, I just got to min the loop bound).

---

(I think you also need to update this line: `return v1.len() * NIBBLE_PER_BYTE;` below to use `cmp::min()` instead of `v1.len()`)